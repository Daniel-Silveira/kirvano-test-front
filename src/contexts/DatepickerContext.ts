import { createContext } from "react";

import { Period, DateValueType, DateType } from "../types";
import dayjs from "dayjs";

interface DatepickerStore {
  period: Period;
  showFooter?: boolean;
  placeholder?: string | null;
  value: DateValueType;
  disabled?: boolean;
  minDate?: DateType | null;
  maxDate?: DateType | null;
  dayHover: string | null;
  toggleVisibilityDatepicker: () => void;
  changePeriod: (period: Period) => void;
  changeDayHover: (day: string | null) => void;
  updateFirstDate: (date: dayjs.Dayjs) => void;
  changeDatepickerValue: (
    value: DateValueType,
    e?: HTMLInputElement | null | undefined
  ) => void;
}

const DatepickerContext = createContext<DatepickerStore>({
  period: { start: null, end: null },
  showFooter: false,
  value: null,
  disabled: false,
  minDate: null,
  maxDate: null,
  dayHover: null,
  changePeriod: (period) => {},
  toggleVisibilityDatepicker: () => {},
  changeDayHover: (day: string | null) => {},
  updateFirstDate: (date) => {},
  changeDatepickerValue: (
    value: DateValueType,
    e: HTMLInputElement | null | undefined
  ) => {},
});

export default DatepickerContext;
