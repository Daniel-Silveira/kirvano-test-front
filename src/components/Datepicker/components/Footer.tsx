import { useContext } from "react";
import { DATE_FORMAT } from "../../../constants";
import { Button } from "../../shared/Button";

import dayjs from "dayjs";
import DatepickerContext from "../../../contexts/DatepickerContext";

const Footer: React.FC = () => {
  const { toggleVisibilityDatepicker, period, changeDatepickerValue } =
    useContext(DatepickerContext);

  return (
    <div className="flex items-center justify-end pb-2.5 pt-3">
      <div className="w-full md:w-auto flex items-center justify-center space-x-3">
        <Button
          onClick={() => {
            toggleVisibilityDatepicker();
          }}
          label="Cancel"
        />
        <Button
          onClick={() => {
            if (period.start && period.end) {
              changeDatepickerValue({
                startDate: dayjs(period.start).format(DATE_FORMAT),
                endDate: dayjs(period.end).format(DATE_FORMAT),
              });
              toggleVisibilityDatepicker();
            }
          }}
          label="Apply"
          primary
        />
      </div>
    </div>
  );
};

export default Footer;
