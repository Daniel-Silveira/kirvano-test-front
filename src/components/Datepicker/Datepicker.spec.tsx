import { fireEvent, render } from "@testing-library/react";
import DatepickerContext from "../../contexts/DatepickerContext";
import Footer from "./components/Footer";

describe("[Datepicker] components", () => {
  describe("Footer component", () => {
    it("should render Cancel and Apply buttons", () => {
      const toggleVisibilityDatepicker = jest.fn();
      const changeDatepickerValue = jest.fn();
      const startDate = "2024-04-01";
      const endDate = "2024-04-15";

      const { getByText, container } = render(
        <DatepickerContext.Provider
          value={{
            toggleVisibilityDatepicker,
            period: { start: startDate, end: endDate },
            changeDatepickerValue,
          }}
        >
          <Footer />
        </DatepickerContext.Provider>
      );

      const cancelButton = getByText("Cancel");
      const applyButton = getByText("Apply");

      expect(cancelButton).toBeInTheDocument();
      expect(applyButton).toBeInTheDocument();

      fireEvent.click(cancelButton);
      expect(toggleVisibilityDatepicker).toHaveBeenCalledTimes(1);

      fireEvent.click(applyButton);
      expect(changeDatepickerValue).toHaveBeenCalledWith({
        startDate,
        endDate,
      });
      expect(toggleVisibilityDatepicker).toHaveBeenCalledTimes(2);
      expect(container).toMatchSnapshot();
    });

    it("should not call changeDatepickerValue if period start or end is missing", () => {
      const toggleVisibilityDatepicker = jest.fn();
      const changeDatepickerValue = jest.fn();

      const { getByText } = render(
        <DatepickerContext.Provider
          value={{
            toggleVisibilityDatepicker,
            period: { start: "", end: "" },
            changeDatepickerValue,
          }}
        >
          <Footer />
        </DatepickerContext.Provider>
      );

      const applyButton = getByText("Apply");

      fireEvent.click(applyButton);
      expect(changeDatepickerValue).not.toHaveBeenCalled();
      expect(toggleVisibilityDatepicker).not.toHaveBeenCalled();
    });
  });
});
