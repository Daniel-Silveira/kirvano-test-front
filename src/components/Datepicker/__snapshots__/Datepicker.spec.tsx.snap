// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Datepicker] components Footer component should render Cancel and Apply buttons 1`] = `
<div>
  <div
    class="flex items-center justify-end pb-2.5 pt-3 border-t border-gray-300 dark:border-gray-700"
  >
    <div
      class="w-full md:w-auto flex items-center justify-center space-x-3"
    >
      <button
        class="bg-transparent text-white font-medium border px-4 py-2 text-sm rounded-md"
      >
        Cancel
      </button>
      <button
        class="bg-primary text-white font-medium border border-primary px-4 py-2 text-sm rounded-md"
      >
        Apply
      </button>
    </div>
  </div>
</div>
`;
